File: solutions/0285. Inorder Successor in BST/001.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> Optional[TreeNode]:
        def inorder(node):
            nonlocal ans
            if node is None or ans: return
            inorder(node.left)
            if node.val >p.val and not ans:
                ans = node
                return
            inorder(node.right)
        ans = None
        inorder(root)
        return ans

File: solutions/1050. Construct Binary Search Tree from Preorder Traversal/001.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
   private TreeNode bstFromPreorder(int[] preorder, int start, int end) {
        if (start < 0 || end >= preorder.length || start > end) return null;
        int val = preorder[start];
        TreeNode root = new TreeNode(val);
        int rstart = start + 1;
        while (rstart <= end && preorder[rstart] < val) rstart++;
        root.left = bstFromPreorder(preorder, start + 1, rstart - 1);
        root.right = bstFromPreorder(preorder, rstart, end);
        return root;
    }

    public TreeNode bstFromPreorder(int[] preorder) {
        return bstFromPreorder(preorder, 0, preorder.length - 1);
    }
}